packages/frontend/src/AfgHandling.ts

```diff
@@ -0,0 +1,265 @@
+    this.initialiseConsensusView(consensusInfo, height, addr, voter);
+
+    const [consensusView, index] = this.getConsensusView(consensusInfo, height);
+
+    consensusView[addr][voter].Finalized = true;
+    consensusView[addr][voter].FinalizedHeight = height;
+    consensusView[addr][voter].ImplicitFinalized = true;
+    consensusView[addr][voter].ImplicitPointer = to;
+
+    // this is extrapolated. if this app was just started up we
+    // might not yet have received prevotes/precommits. but
+    // those are a necessary precondition for finalization, so
+    // we can set them and display them in the ui.
+    consensusView[addr][voter].Prevote = true;
+    consensusView[addr][voter].Precommit = true;
+    consensusView[addr][voter].ImplicitPrevote = true;
+    consensusView[addr][voter].ImplicitPrecommit = true;
+
+    if (index !== -1) {
+      consensusInfo[index] = [height, consensusView];
+    } else {
+      consensusInfo.unshift([height, consensusView]);
+    }
+
+    this.updateState({consensusInfo});
+  }
+
+  // Initializes the `ConsensusView` with empty objects.
+  private initialiseConsensusView(
+    consensusInfo: Types.ConsensusInfo,
+    height: Types.BlockNumber,
+    own: Types.Address,
+    other: Types.Address,
+  ) {
+    const index =
+      consensusInfo.findIndex(([blockNumber,]) => blockNumber === height);
+
+    let consensusView;
+    if (index === -1) {
+      consensusView = {} as Types.ConsensusView;
+    } else {
+      const found = consensusInfo[index];
+      const [, foundView] = found;
+      consensusView = foundView;
+    }
+
+    if (!consensusView[own]) {
+      consensusView[own] = {} as Types.ConsensusState;
+    }
+
+    if (!consensusView[own][other]) {
+      consensusView[own][other] = {} as Types.ConsensusDetail;
+    }
+
+    if (index === -1) {
+      // append at the beginning
+      consensusInfo.unshift([height, consensusView]);
+    } else {
+      consensusInfo[index] = [height, consensusView];
+    }
+  }
+
+  // Fill the block cache back from the `to` number to the last block.
+  // The function `f` is used to evaluate if we should continue backfilling.
+  // `f` returns false when backfilling the cache should be stopped, true to continue.
+  //
+  // Returns block number until which we backfilled.
+  private backfill(
+    consensusInfo: Types.ConsensusInfo,
+    to: Types.BlockNumber,
+    f: Maybe<(i: Types.BlockNumber, consensusView: Types.ConsensusView) => boolean>,
+    own: Types.Address,
+    other: Types.Address,
+  ): Types.BlockNumber {
+    let cont = true;
+    for (const i of consensusInfo) {
+      const [height,] = i;
+      if (height >= to) {
+        continue;
+      }
+
+      this.initialiseConsensusView(consensusInfo, height, own, other);
+
+      const index =
+        consensusInfo.findIndex(([blockNumber,]) => blockNumber === height);
+      if (index === -1) {
+        cont = false;
+        break;
+      }
```

Comment:
Shorter way to write the whole loop (the index check in the middle is superficial as noted above):

```ts
for (const [height, consensusView] of consensusInfo) {
  if (height >= to) {
    contine;
  }

  this.initialiseConsensusView(consensusInfo, height, own, other);

  const cont = f ? f(height, consensusView) : true;
  if (!cont) {
    break;
  }
}
```

It might make sense to extract a function that builds up `Types.ConsensusView` that doesn't need to index into `consensusInfo`, that is:

```ts
this.initialiseConsensusView(consensusInfo, height, own, other);
```

Could become:
```ts
this.intialiseConsensusViewByRef(consensusView, own, other);
```

The former can then use the latter internally to keep things DRY.